health/polling.go

```diff
@@ -157,35 +157,38 @@ func (r *pollingChecker) New(
 		defer ticker.Stop()
 
 		for {
-			ctx, cancel := context.WithTimeout(ctx, r.timeout)
-			defer cancel()
-
-			result := r.prober.Probe(ctx, conn)
-
-			lastState := state
-			switch {
-			case result == StateHealthy && (state == StateUnhealthy || state == StateDegraded || state == StateUnknown):
-				counter++
-				if counter >= r.healthyThreshold {
+			func() {
+				ctx, cancel := context.WithTimeout(ctx, r.timeout)
+				defer cancel()
+
+				result := r.prober.Probe(ctx, conn)
+
+				lastState := state
+
+				switch {
+				case result == StateHealthy && (state == StateUnhealthy || state == StateDegraded || state == StateUnknown):
+					counter++
+					if counter >= r.healthyThreshold {
+						state = result
+						counter = 0
+					}
+
+				case state == StateHealthy && (result == StateUnhealthy || result == StateDegraded || result == StateUnknown):
+					counter++
+					if counter >= r.unhealthyThreshold {
+						state = result
+						counter = 0
+					}
+
+				default:
 					state = result
 					counter = 0
 				}
 
-			case state == StateHealthy && (result == StateUnhealthy || result == StateDegraded || result == StateUnknown):
-				counter++
-				if counter >= r.unhealthyThreshold {
-					state = result
-					counter = 0
+				if lastState != state {
+					tracker.UpdateHealthState(conn, result)
 				}
-
-			default:
-				state = result
-				counter = 0
-			}
-
-			if lastState != state {
-				tracker.UpdateHealthState(conn, result)
-			}
+			}()
 
 			select {
 			case <-ctx.Done():
```

Comment:
This `return` used to trigger when `ctx` created for the probe is done, meaning the loop ends. Because `defer cancel()` was essentially deferring accumulated closures to the end of the loop itself, it didn't run into a "probe-only-once" situation, but a timeout would similarly cause the loop to break prematurely.

Now, this should only be triggered by closing the poller / http client, not by timeout.